File: packages/typescript/core/src/utils/validation.ts
/**
 * Configuration validation utilities that delegate to native Rust validators via FFI.
 *
 * These functions call the native validation functions exposed through NAPI-RS,
 * ensuring a single source of truth for validation logic across all language bindings.
 *
 * NOTE: These functions are bound at runtime to the native NAPI-RS module.
 * The actual validation logic lives in crates/kreuzberg-ffi/src/validation.rs
 * and is exposed through crates/kreuzberg-node/src/lib.rs
 */

let nativeModule: Record<string, any>;

function getNativeModule(): Record<string, any> {
	if (!nativeModule) {
		try {
			nativeModule = require("kreuzberg-node");
		} catch (error) {
			throw new Error("Unable to load native kreuzberg-node module. Please ensure it is properly compiled.");
		}
	}
	return nativeModule;
}

/**
 * Validates a binarization method string.
 *
 * Valid methods: "otsu", "adaptive", "sauvola"
 *
 * @param method The binarization method to validate
 * @throws if the method is invalid
 *
 * @example
 * ```typescript
 * import { validateBinarizationMethod } from '@kreuzberg/core';
 *
 * try {
 *   validateBinarizationMethod('otsu');
 *   console.log('Valid method');
 * } catch (error) {
 *   console.error('Invalid method:', error.message);
 * }
 * ```
 */
export function validateBinarizationMethod(method: string): void {
	const validator = getNativeModule().validateBinarizationMethod;
	if (!validator(method)) {
		throw new Error(`Invalid binarization method: ${method}`);
	}
}

/**
 * Validates an OCR backend string.
 *
 * Valid backends: "tesseract", "easyocr", "paddleocr"
 *
 * @param backend The OCR backend to validate
 * @throws if the backend is invalid
 *
 * @example
 * ```typescript
 * import { validateOcrBackend } from '@kreuzberg/core';
 *
 * try {
 *   validateOcrBackend('tesseract');
 * } catch (error) {
 *   console.error('Invalid backend:', error.message);
 * }
 * ```
 */
export function validateOcrBackend(backend: string): void {
	const validator = getNativeModule().validateOcrBackend;
	if (!validator(backend)) {
		throw new Error(`Invalid OCR backend: ${backend}`);
	}
}

/**
 * Validates a language code (ISO 639-1 or 639-3 format).
 *
 * Accepts both 2-letter codes (e.g., "en", "de") and 3-letter codes (e.g., "eng", "deu").
 *
 * @param code The language code to validate
 * @throws if the code is invalid
 *
 * @example
 * ```typescript
 * import { validateLanguageCode } from '@kreuzberg/core';
 *
 * try {
 *   validateLanguageCode('en');
 * } catch (error) {
 *   console.error('Invalid language code:', error.message);
 * }
 * ```
 */
export function validateLanguageCode(code: string): void {
	const validator = getNativeModule().validateLanguageCode;
	if (!validator(code)) {
		throw new Error(`Invalid language code: ${code}`);
	}
}

/**
 * Validates a token reduction level string.
 *
 * Valid levels: "off", "light", "moderate", "aggressive", "maximum"
 *
 * @param level The token reduction level to validate
 * @throws if the level is invalid
 *
 * @example
 * ```typescript
 * import { validateTokenReductionLevel } from '@kreuzberg/core';
 *
 * try {
 *   validateTokenReductionLevel('moderate');
 * } catch (error) {
 *   console.error('Invalid token reduction level:', error.message);
 * }
 * ```
 */
export function validateTokenReductionLevel(level: string): void {
	const validator = getNativeModule().validateTokenReductionLevel;
	if (!validator(level)) {
		throw new Error(`Invalid token reduction level: ${level}`);
	}
}

/**
 * Validates a Tesseract Page Segmentation Mode (PSM) value.
 *
 * Valid range: 0-13
 *
 * @param psm The PSM value to validate
 * @throws if the PSM is invalid
 *
 * @example
 * ```typescript
 * import { validateTesseractPsm } from '@kreuzberg/core';
 *
 * try {
 *   validateTesseractPsm(3);
 * } catch (error) {
 *   console.error('Invalid PSM:', error.message);
 * }
 * ```
 */
export function validateTesseractPsm(psm: number): void {
	const validator = getNativeModule().validateTesseractPsm;
	if (!validator(psm)) {
		throw new Error(`Invalid Tesseract PSM: ${psm}. Valid range: 0-13`);
	}
}

/**
 * Validates a Tesseract OCR Engine Mode (OEM) value.
 *
 * Valid range: 0-3
 *
 * @param oem The OEM value to validate
 * @throws if the OEM is invalid
 *
 * @example
 * ```typescript
 * import { validateTesseractOem } from '@kreuzberg/core';
 *
 * try {
 *   validateTesseractOem(1);
 * } catch (error) {
 *   console.error('Invalid OEM:', error.message);
 * }
 * ```
 */
export function validateTesseractOem(oem: number): void {
	const validator = getNativeModule().validateTesseractOem;
	if (!validator(oem)) {
		throw new Error(`Invalid Tesseract OEM: ${oem}. Valid range: 0-3`);
	}
}

/**
 * Validates a tesseract output format string.
 *
 * Valid formats: "text", "markdown"
 *
 * @param format The output format to validate
 * @throws if the format is invalid
 *
 * @example
 * ```typescript
 * import { validateOutputFormat } from '@kreuzberg/core';
 *
 * try {
 *   validateOutputFormat('markdown');
 * } catch (error) {
 *   console.error('Invalid output format:', error.message);
 * }
 * ```
 */
export function validateOutputFormat(format: string): void {
	const validator = getNativeModule().validateOutputFormat;
	if (!validator(format)) {
		throw new Error(`Invalid output format: ${format}`);
	}
}

/**
 * Validates a confidence threshold value.
 *
 * Valid range: 0.0 to 1.0 (inclusive)
 *
 * @param confidence The confidence threshold to validate
 * @throws if the confidence is invalid
 *
 * @example
 * ```typescript
 * import { validateConfidence } from '@kreuzberg/core';
 *
 * try {
 *   validateConfidence(0.75);
 * } catch (error) {
 *   console.error('Invalid confidence:', error.message);
 * }
 * ```
 */
export function validateConfidence(confidence: number): void {
	const validator = getNativeModule().validateConfidence;
	if (!validator(confidence)) {
		throw new Error(`Invalid confidence: ${confidence}. Valid range: 0.0-1.0`);
	}
}

/**
 * Validates a DPI (dots per inch) value.
 *
 * Valid range: 1-2400
 *
 * @param dpi The DPI value to validate
 * @throws if the DPI is invalid
 *
 * @example
 * ```typescript
 * import { validateDpi } from '@kreuzberg/core';
 *
 * try {
 *   validateDpi(300);
 * } catch (error) {
 *   console.error('Invalid DPI:', error.message);
 * }
 * ```
 */
export function validateDpi(dpi: number): void {
	const validator = getNativeModule().validateDpi;
	if (!validator(dpi)) {
		throw new Error(`Invalid DPI: ${dpi}. Valid range: 1-2400`);
	}
}

/**
 * Validates chunking parameters.
 *
 * Checks that `maxChars > 0` and `maxOverlap < maxChars`.
 *
 * @param maxChars Maximum characters per chunk
 * @param maxOverlap Maximum overlap between chunks
 * @throws if the parameters are invalid
 *
 * @example
 * ```typescript
 * import { validateChunkingParams } from '@kreuzberg/core';
 *
 * try {
 *   validateChunkingParams(1000, 200);
 * } catch (error) {
 *   console.error('Invalid chunking params:', error.message);
 * }
 * ```
 */
export function validateChunkingParams(maxChars: number, maxOverlap: number): void {
	const validator = getNativeModule().validateChunkingParams;
	if (!validator(maxChars, maxOverlap)) {
		throw new Error(`Invalid chunking params: maxChars=${maxChars}, maxOverlap=${maxOverlap}`);
	}
}

/**
 * Get all valid binarization methods.
 *
 * @returns Array of valid binarization methods
 *
 * @example
 * ```typescript
 * import { getValidBinarizationMethods } from '@kreuzberg/core';
 *
 * const methods = await getValidBinarizationMethods();
 * console.log(methods); // ['otsu', 'adaptive', 'sauvola']
 * ```
 */
export async function getValidBinarizationMethods(): Promise<string[]> {
	const getter = getNativeModule().getValidBinarizationMethods;
	return getter();
}

/**
 * Get all valid language codes.
 *
 * @returns Array of valid language codes (both 2-letter and 3-letter codes)
 *
 * @example
 * ```typescript
 * import { getValidLanguageCodes } from '@kreuzberg/core';
 *
 * const codes = await getValidLanguageCodes();
 * console.log(codes); // ['en', 'de', 'fr', ..., 'eng', 'deu', 'fra', ...]
 * ```
 */
export async function getValidLanguageCodes(): Promise<string[]> {
	const getter = getNativeModule().getValidLanguageCodes;
	return getter();
}

/**
 * Get all valid OCR backends.
 *
 * @returns Array of valid OCR backends
 *
 * @example
 * ```typescript
 * import { getValidOcrBackends } from '@kreuzberg/core';
 *
 * const backends = await getValidOcrBackends();
 * console.log(backends); // ['tesseract', 'easyocr', 'paddleocr']
 * ```
 */
export async function getValidOcrBackends(): Promise<string[]> {
	const getter = getNativeModule().getValidOcrBackends;
	return getter();
}

/**
 * Get all valid token reduction levels.
 *
 * @returns Array of valid token reduction levels
 *
 * @example
 * ```typescript
 * import { getValidTokenReductionLevels } from '@kreuzberg/core';
 *
 * const levels = await getValidTokenReductionLevels();
 * console.log(levels); // ['off', 'light', 'moderate', 'aggressive', 'maximum']
 * ```
 */
export async function getValidTokenReductionLevels(): Promise<string[]> {
	const getter = getNativeModule().getValidTokenReductionLevels;
	return getter();
}

File: crates/kreuzberg-php/src/validation.rs
//! Validation functions for PHP bindings
//!
//! Provides validation functions for configuration parameters.

use ext_php_rs::prelude::*;

/// Validate a binarization method.
///
/// # Parameters
///
/// - `method` (string): The binarization method to validate
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Valid Methods
///
/// - "otsu"
/// - "adaptive"
/// - "sauvola"
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_binarization_method("otsu")) {
///     echo "Valid method\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_binarization_method(method: String) -> bool {
    kreuzberg::core::config_validation::validate_binarization_method(&method).is_ok()
}

/// Validate an OCR backend.
///
/// # Parameters
///
/// - `backend` (string): The OCR backend to validate
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Valid Backends
///
/// - "tesseract"
/// - "easyocr"
/// - "paddleocr"
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_ocr_backend("tesseract")) {
///     echo "Valid backend\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_ocr_backend(backend: String) -> bool {
    kreuzberg::core::config_validation::validate_ocr_backend(&backend).is_ok()
}

/// Validate a language code (ISO 639-1 or 639-3 format).
///
/// # Parameters
///
/// - `code` (string): The language code to validate
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_language_code("eng")) {
///     echo "Valid language code\n";
/// }
/// if (kreuzberg_validate_language_code("en")) {
///     echo "Also valid (2-letter code)\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_language_code(code: String) -> bool {
    kreuzberg::core::config_validation::validate_language_code(&code).is_ok()
}

/// Validate a token reduction level.
///
/// # Parameters
///
/// - `level` (string): The token reduction level to validate
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Valid Levels
///
/// - "off"
/// - "light"
/// - "moderate"
/// - "aggressive"
/// - "maximum"
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_token_reduction_level("moderate")) {
///     echo "Valid level\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_token_reduction_level(level: String) -> bool {
    kreuzberg::core::config_validation::validate_token_reduction_level(&level).is_ok()
}

/// Validate a Tesseract Page Segmentation Mode (PSM) value.
///
/// # Parameters
///
/// - `psm` (int): The PSM value to validate (valid range: 0-13)
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_tesseract_psm(3)) {
///     echo "Valid PSM\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_tesseract_psm(psm: i64) -> bool {
    kreuzberg::core::config_validation::validate_tesseract_psm(psm as i32).is_ok()
}

/// Validate a Tesseract OCR Engine Mode (OEM) value.
///
/// # Parameters
///
/// - `oem` (int): The OEM value to validate (valid range: 0-3)
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_tesseract_oem(3)) {
///     echo "Valid OEM\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_tesseract_oem(oem: i64) -> bool {
    kreuzberg::core::config_validation::validate_tesseract_oem(oem as i32).is_ok()
}

/// Validate a Tesseract output format string.
///
/// # Parameters
///
/// - `format` (string): The output format to validate
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Valid Formats
///
/// - "text"
/// - "markdown"
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_output_format("markdown")) {
///     echo "Valid format\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_output_format(format: String) -> bool {
    kreuzberg::core::config_validation::validate_output_format(&format).is_ok()
}

/// Validate a confidence threshold value.
///
/// # Parameters
///
/// - `confidence` (float): The confidence threshold to validate (valid range: 0.0-1.0)
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_confidence(0.8)) {
///     echo "Valid confidence\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_confidence(confidence: f64) -> bool {
    kreuzberg::core::config_validation::validate_confidence(confidence).is_ok()
}

/// Validate a DPI (dots per inch) value.
///
/// # Parameters
///
/// - `dpi` (int): The DPI value to validate (must be positive, typically 72-600)
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_dpi(300)) {
///     echo "Valid DPI\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_dpi(dpi: i64) -> bool {
    kreuzberg::core::config_validation::validate_dpi(dpi as i32).is_ok()
}

/// Validate chunking parameters.
///
/// # Parameters
///
/// - `max_chars` (int): Maximum characters per chunk (must be > 0)
/// - `max_overlap` (int): Maximum overlap between chunks (must be < max_chars)
///
/// # Returns
///
/// `true` if valid, `false` otherwise
///
/// # Example
///
/// ```php
/// if (kreuzberg_validate_chunking_params(1000, 200)) {
///     echo "Valid chunking params\n";
/// }
/// ```
#[php_function]
pub fn kreuzberg_validate_chunking_params(max_chars: i64, max_overlap: i64) -> bool {
    kreuzberg::core::config_validation::validate_chunking_params(max_chars as usize, max_overlap as usize).is_ok()
}

/// Get list of valid binarization methods.
///
/// # Returns
///
/// Array of valid binarization method names
///
/// # Example
///
/// ```php
/// $methods = kreuzberg_get_valid_binarization_methods();
/// print_r($methods); // ["otsu", "adaptive", "sauvola"]
/// ```
#[php_function]
pub fn kreuzberg_get_valid_binarization_methods() -> Vec<String> {
    vec!["otsu".to_string(), "adaptive".to_string(), "sauvola".to_string()]
}

/// Get list of valid language codes.
///
/// Returns both 2-letter (ISO 639-1) and 3-letter (ISO 639-3) language codes.
///
/// # Returns
///
/// Array of valid language codes
///
/// # Example
///
/// ```php
/// $codes = kreuzberg_get_valid_language_codes();
/// // Includes: "en", "eng", "de", "deu", etc.
/// ```
#[php_function]
pub fn kreuzberg_get_valid_language_codes() -> Vec<String> {
    vec![
        "en", "de", "fr", "es", "it", "pt", "nl", "pl", "ru", "zh", "ja", "ko", "bg", "cs", "da", "el", "et", "fi",
        "hu", "lt", "lv", "ro", "sk", "sl", "sv", "uk", "ar", "hi", "th", "tr", "vi", "eng", "deu", "fra", "spa",
        "ita", "por", "nld", "pol", "rus", "zho", "jpn", "kor", "ces", "dan", "ell", "est", "fin", "hun", "lit", "lav",
        "ron", "slk", "slv", "swe", "tur",
    ]
    .iter()
    .map(|s| s.to_string())
    .collect()
}

/// Get list of valid OCR backends.
///
/// # Returns
///
/// Array of valid OCR backend names
///
/// # Example
///
/// ```php
/// $backends = kreuzberg_get_valid_ocr_backends();
/// print_r($backends); // ["tesseract", "easyocr", "paddleocr"]
/// ```
#[php_function]
pub fn kreuzberg_get_valid_ocr_backends() -> Vec<String> {
    vec!["tesseract".to_string(), "easyocr".to_string(), "paddleocr".to_string()]
}

/// Get list of valid token reduction levels.
///
/// # Returns
///
/// Array of valid token reduction level names
///
/// # Example
///
/// ```php
/// $levels = kreuzberg_get_valid_token_reduction_levels();
/// print_r($levels); // ["off", "light", "moderate", "aggressive", "maximum"]
/// ```
#[php_function]
pub fn kreuzberg_get_valid_token_reduction_levels() -> Vec<String> {
    vec![
        "off".to_string(),
        "light".to_string(),
        "moderate".to_string(),
        "aggressive".to_string(),
        "maximum".to_string(),
    ]
}
